src/cpuid/src/intel/mod.rs

```diff
@@ -0,0 +1,439 @@
+
+/// A structure matching supporting the  Intel CPUID specification as described in
+/// [Intel® 64 and IA-32 Architectures Software Developer's Manual Combined Volumes 2A, 2B, 2C, and 2D: Instruction Set Reference, A-Z](https://cdrdv2.intel.com/v1/dl/getContent/671110)
+/// .
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct IntelCpuid(pub std::collections::BTreeMap<CpuidKey, CpuidEntry>);
+
+impl CpuidTrait for IntelCpuid {
+    /// Gets a given sub-leaf.
+    #[inline]
+    fn get(&self, key: &CpuidKey) -> Option<&CpuidEntry> {
+        self.0.get(key)
+    }
+
+    /// Gets a given sub-leaf.
+    #[inline]
+    fn get_mut(&mut self, key: &CpuidKey) -> Option<&mut CpuidEntry> {
+        self.0.get_mut(key)
+    }
+}
+
+/// Error type for [`IntelCpuid::default_brand_string`].
+#[cfg(cpuid)]
+#[derive(Debug, Eq, PartialEq, thiserror::Error)]
+pub enum DefaultBrandStringError {
+    /// Missing frequency.
+    #[error("Missing frequency: {0:?}.")]
+    Missingfrequency([u8; crate::BRAND_STRING_LENGTH]),
+    /// Missing space.
+    #[error("Missing space: {0:?}.")]
+    MissingSpace([u8; crate::BRAND_STRING_LENGTH]),
+    /// Insufficient space in brand string.
+    #[error("Insufficient space in brand string.")]
+    Overflow,
+}
+
+impl IntelCpuid {
+    /// Gets the brand string always used for Intel.
+    ///
+    /// # Errors
+    ///
+    /// When unable to parse the host brand string.
+    /// `brand_string.try_into().unwrap()` cannot panic as we know
+    /// `brand_string.len() == BRAND_STRING_LENGTH`
+    ///
+    /// # Panics
+    ///
+    /// Never.
+    // As we pass through host frequency, we require CPUID and thus `cfg(cpuid)`.
+    // TODO: Use `split_array_ref`
+    // (https://github.com/firecracker-microvm/firecracker/issues/3347)
+    #[allow(
+        clippy::indexing_slicing,
+        clippy::integer_arithmetic,
+        clippy::arithmetic_side_effects
+    )]
+    #[cfg(cpuid)]
+    #[inline]
+    pub fn default_brand_string(
+    ) -> Result<[u8; super::BRAND_STRING_LENGTH], DefaultBrandStringError> {
+        /// We always use this brand string.
+        const DEFAULT_BRAND_STRING_BASE: &[u8] = b"Intel(R) Xeon(R) Processor @";
+
+        // Get host brand string.
+        // This will look like b"Intel(4) Xeon(R) Processor @ 3.00GHz".
+        let host_brand_string: [u8; super::BRAND_STRING_LENGTH] = super::host_brand_string();
+
+        // The slice of the host string before the frequency suffix
+        // e.g. b"Intel(4) Xeon(R) Processor @ 3.00" and "GHz"
+        let (before, after) = 'outer: {
+            for i in 0..host_brand_string.len() {
+                // Find position of b"THz" or b"GHz" or b"MHz"
+                if let [b'T' | b'G' | b'M', b'H', b'z', ..] = host_brand_string[i..] {
+                    break 'outer Ok(host_brand_string.split_at(i));
+                }
+            }
+            Err(DefaultBrandStringError::Missingfrequency(host_brand_string))
+        }?;
+        debug_assert_eq!(
```

Comment:
Many of these assertions would be replaced with comments, that make the same points. I think in this way they act as self testing comments.

In many of these instances the only way to test is with assertions. Given the code:
```rust
debug_assert_eq!(brand_string.len(), super::BRAND_STRING_LENGTH);
// SAFETY: Padding ensures `brand_string.len() == BRAND_STRING_LENGTH`.
Ok(unsafe { brand_string.try_into().unwrap_unchecked() })
```
We cannot test for the case when `brand_string.len() != super::BRAND_STRING_LENGTH` without this assertion since when the function calls `.unwrap_unchecked()` on an `Err` it will result in undefined behavior.

Notably if we look at the standard library definition of `unwrap_unchecked()` it also uses a `debug_assert`:
```rust
    pub const unsafe fn unwrap_unchecked(self) -> T {
        debug_assert!(self.is_some());
        match self {
            Some(val) => val,
            // SAFETY: the safety contract must be upheld by the caller.
            None => unsafe { hint::unreachable_unchecked() },
        }
    }
```
unfortunately since we use the standard library pre-built in release mode this assertion is not included, so to test this it may be reasonable to ensure it is safe with our own assertion.

Reply:
1. I do not agree that self testing comments should stay in prod code (doing this as a practice would result in hard to maintain code).
2. In this specific `debug_assert_eq`, checking that `BRAND_STRING_LENGTH` is equal to `before.len().checked_add(after.len())` tests  the `split_at` Rust function; we test that functionality has not changed between Rust versions with unit and integration tests.

Reply:
>1. I do not agree that self testing comments should stay in prod code (doing this as a practice would result in hard to maintain code).

These asserts are not compiled into the release binary so would not be present in production.

Reply:
I've removed the asserts.

Reply:
I have removed them.